refactor: improve linter and formatter intros (#142)
biomejs/website: src/content/docs/linter/index.mdx

---

import NumberOfRules from "@/components/generated/NumberOfRules.astro";
import PackageManagerBiomeCommand from "@/components/PackageManagerBiomeCommand.astro";

> Biome's linter statically analyzes your code to find and fix common errors and to help you write better code.

Reviewer: ```suggestion
Biome's linter statically analyzes your code to find and fix common errors and to help you write better, modern code.
```

It [supports multiple languages](/internals/language-support) and provides a total of [**<NumberOfRules/> rules**](/linter/rules/).

## CLI

The following command runs the linter on all files in the `src` directory:

<PackageManagerBiomeCommand command="lint ./src" />

> Note that if you pass a glob as a parameter, it will be expanded by your shell.
> The result of the expansion depends on your shell and its configuration.
> For example, some shells don't support the recursive glob `**` or the alternation `{}` in the following command:

Reviewer: ```suggestion
Note that if you pass a glob as a parameter, your shell will expand it.
The result of the expansion depends on your shell and its configuration.
For example, some shells don't support the recursive glob `**` or the alternation `{}` in the following command:
```

We have the same phrase in the formatter. Maybe we should have it in a common section, otherwise we will struggle to update both.

Author: I will try to address the duplication concern in a future PR.

```shell
biome lint ./src/**/*.test.{js,ts}
```

For more information about all the available options, check the [CLI reference](/reference/cli#biome-lint).

## Rules

The linter is organized into rules.

> A rule emit a diagnostic when it encounters a code that doesn't meet its requirements.
> For example, the [noDebugger](http://localhost:4321/linter/rules/no-debugger) rule reports the use of the `debugger` instruction in JavaScript code.

Reviewer: ```suggestion
A rule emits a diagnostic when it encounters a code that doesn't meet its requirements.
For example, the [noDebugger](http://localhost:4321/linter/rules/no-debugger) rule reports using the `debugger` instruction in JavaScript code.
```

> A rule emits diagnostics with a `warn` or `error` severity.
> Diagnostics with an `error` severity cause the command to exit with a non-zero code,
> While diagnostics with a `warn` severity don't cause the command to fail.
> You can turn warnings into errors `--error-on-warnings` CLI option:

Reviewer: I would change the explanation and focus more on the rules rather than discussing diagnostics. In the future, a full, in-depth explanation of our diagnostics should be on a separate page.

My suggestion:

```suggestion
By default, **recommended** rules emit a diagnostic with `error` severity and will cause the command to fail. Via configuration, you can demote recommended rules to emit warnings or promote non-recommended rules to emit errors.

You can turn warnings into errors `--error-on-warnings` CLI option:
```

```shell
biome lint --error-on-warnings ./src
```

By default, the Biome linter only runs [**recommended rules**](/linter/rules/).
Recommended rules emit diagnostics with the `error` severity.

Rules are organized into groups.
For example, the `noDebugger` rule is part of the [`suspicious` group](/linter/rules#suspicious).

> A Rule from this group detects code that is likely to be incorrect or useless.
> The description of each group can be found on the [rules page](/linter/rules/).

Reviewer: ```suggestion
Our [rules page](/linter/rules/) explains each group in-depth.
```

There is no need to explain a group if there's a page already for that (we avoid repeating information and causing too much fragmentation).

Unlike other linters, we don't provide any rules that check for code formatting.
This kind of checking is covered by our [code formatter](/formatter/).

Many rules provide a **code fix** that can be automatically applied.

> Biome distinguishes between safe and unsafe code fixes.

Reviewer: ```suggestion
Biome distinguishes between **safe** and **unsafe** code fixes.
```

<PackageManagerBiomeCommand command="lint --apply ./src" />

### Unsafe fixes

Unsafe fixes may change the semantic of your program.
Therefore, it's advised to manually review the changes.

To apply _unsafe fixes_, use `--apply-unsafe`:

<PackageManagerBiomeCommand command="lint --apply-unsafe ./src" />

Note that this command also applies safe fixes.

### Rule pillars

We believe that rules should be informative and explain to the user why a rule is triggered and tell the user what they should to do fix the error.
A rule should follow these **pillars**:

1. Explain to the user the error. Generally, this is the message of the diagnostic.
2. Explain to the user **why** the error is triggered. Generally, this is implemented with an additional node.
3. Tell the user what they should do. Generally, this is implemented using a code action.
   If a code action is not applicable a note should tell the user what they should do to fix the error.

If you think a rule doesn't follow these pillars,
please [open an issue](https://github.com/biomejs/biome/issues/new?assignees=&labels=S-To+triage&projects=&template=01_bug.yml&title=%F0%9F%90%9B+%3CTITLE%3E).

## Configuration

> Recommended rules are enabled by default and emit diagnostics with the `error` severity.

Reviewer: ```suggestion
```

Repeated phrase

> Rules that are not recommended are disabled by default, but they can be enabled via [configuration](/guides/how-biome-works#configuration).

Reviewer: If you accept my previous suggestion, this is also a repetition.

> A rule is enabled whether its severity is either `error` or `warn` and is disabled if its severity is `off`.

Reviewer: In our configuration, we have `level`, but here you say, "Its severity is off." I believe this is confusing. I think we shouldn't say "severity", and just say "you can turn off a rule with `"off"`"

          }
        }
      }
    }
  }
}
```

## Ignoring code

There are times when a developer wants to ignore a lint rule for a specific line of the code.
You can achieve this by adding a suppression comment above the line that emits the lint diagnostic.

Suppression comments have the following format:

```js
// biome-ignore lint: <explanation>
// biome-ignore lint/suspicious/noDebugger: <explanation>
```

Where

- `biome-ignore` is the start of a suppression comment;
- `lint` suppresses the linter;
- `/suspicious/noDebugger`: **optional**, group and name of the rule you want to suppress;
- `<explanation>` explanation why the rule is disabled

Here's an example:

```ts
// biome-ignore lint: reason
debugger;
// biome-ignore lint/suspicious/noDebugger: reason
debugger;
```

## Ignoring files

Biome doesn't provide ignore comments that ignore an entire file.
However, you can ignore a file using the Biome configuration file.

You can ignore files for all tools, including the linter,
using the [`files.ignore` configuration](/guides/how-biome-works#include-and-ignore-explained).
By default, Biome ignores the [protected files](/guides/how-biome-works#protected-files).

> If you want that only the linter to ignore some files, you can use `linter.ignore`:

Reviewer: A lot of repeated words equals the formatter. I am worried that it could take too much work if a tomorrow will change things, for example, around how we ignore files. 

We should try to keep things in one place, and use links to redirect users.

Author: I will try to address the duplication concern in a future PR.
